doc/manpages/ccndstart.1
'\" t
.\"     Title: ccndstart
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNDSTART" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccndstart \- Start ccnd in the background and set up forwarding according to configuration
.SH "SYNOPSIS"
.sp
\fBccndstart\fR
.SH "DESCRIPTION"
.sp
The \fBccndstart\fR utility starts ccnd in the background\&. \fBccndstart\fR looks for configuration of forwarding information in \fI$HOME/\&.ccnx/ccnd\&.conf\fR; refer to \fBccndc(1)\fR for the format\&. If \fBccnd\fR is already running, it will be shut down and restarted; any manually\-configured forwarding (with \fBccndc\fR) will need to be re\-executed\&.
.sp
\fBccndstart\fR will start using default values for \fICCND_CAP\fR, \fICCND_DEBUG\fR and the default port 9695 for \fICCN_LOCAL_PORT\fR\&. These environment values may be set to change the cache size, \fBccnd\fR debug logging, and port number, respectively\&. The values may be set in \fI$HOME/\&.ccnx/ccndrc\fR, which is sourced at startup\&.
.SH "OPTIONS"
.sp
This utility does not have have flags or arguments\&.
.sp
These environment variables are handled by \fBccndstart\fR:
.sp
.if n \{\
.RS 4
.\}
.nf
CCND_LOG
CCND_PRELOAD
.fi
.if n \{\
.RE
.\}
.sp
\fICCND_LOG\fR is the name of the log file\&. If none is specified, the log output appears on \fBstderr\fR\&.
.sp
\fICCND_PRELOAD\fR is a list of file names for files containing ccnb\-encoded content objects\&. These will be preloaded into the ccnd\(cqs cache at startup\&. See \fBccnrm(1)\fR for one way to make such a file\&.
.sp
If the environment variable \fICCNR_GLOBAL_PREFIX\fR is set, \fBccndstart\fR will also start an instance of the \fBccnr\fR repository, using the directory \fI$HOME/\&.ccnx/repository\fR to contain the backing files\&. Additional configuration for this instance may be placed in the file \fI$HOME/\&.ccnx/repository/config\fR\&.
.sp
Please see \fBccnd(1)\fR and \fBccnr(1)\fR for additional environment variables\&.
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure
.RE
.SH "FILES"
.sp
.if n \{\
.RS 4
.\}
.nf
$HOME/\&.ccnx/ccnd\&.conf
$HOME/\&.ccnx/ccndrc
.fi
.if n \{\
.RE
.\}
.SH "AUTHOR"
.sp
Michael Plass


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

csrc/util/ccndstart.sh
# Source file: util/ccndstart.sh
# Start ccnd in the background and set up forwarding according to configuration
# 
# Part of the CCNx distribution.
#
# Copyright (C) 2009-2013 Palo Alto Research Center, Inc.
#
# This work is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License version 2 as published by the
# Free Software Foundation.
# This work is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.
#

# This script should be installed in the same place as ccnd, ccndc, ccndsmoketest, ...
# adjust the path to get consistency.
D=`dirname "$0"`
export PATH="$D:$PATH"

# Source a file containing settings, if present.
# To learn about things that you can set, use this command: ccnd -h
test -f $HOME/.ccnx/ccndrc && . $HOME/.ccnx/ccndrc

StuffPreload () {
    # Stuff preloaded content objects into ccnd
    # To use this feature, set CCND_PRELOAD in ~/.ccnx/ccndrc
    # Also has side effect of waiting until ccnd is up, even if no preload
    # First a brief delay to give forked ccnd a chance to start
    ccndsmoketest -u localhost -t 50 recv >/dev/null
    ccndsmoketest -b `for i in $CCND_PRELOAD; do echo send $i; done` >/dev/null
}

# Provide defaults
: ${CCND_CAP:=50000}
: ${CCND_DEBUG:=''}
export CCN_LOCAL_PORT CCND_CAP CCND_DEBUG CCND_AUTOREG CCND_LISTEN_ON CCND_MTU
# The following are rarely used, but include them for completeness
export CCN_LOCAL_SOCKNAME CCND_DATA_PAUSE_MICROSEC CCND_KEYSTORE_DIRECTORY
export CCND_DEFAULT_TIME_TO_STALE CCND_MAX_TIME_TO_STALE CCND_PREFIX
export CCND_MAX_RTE_MICROSEC

# If a ccnd is already running, try to shut it down cleanly.
ccndsmoketest kill 2>/dev/null

# Fork ccnd, with a log file if requested.
if [ "$CCND_LOG" = "" ]
then
	ccnd &
        StuffPreload
else
	: >"$CCND_LOG" || exit 1
	ccnd 2>"$CCND_LOG" &
        StuffPreload 2> /dev/null
fi

# Run ccndc if a static config file is present.
test -f $HOME/.ccnx/ccnd.conf && ccndc -f $HOME/.ccnx/ccnd.conf

# Start a repository as well, if a global prefix has been configured.
if [ "$CCNR_GLOBAL_PREFIX" != "" ]
then
mkdir -p $HOME/.ccnx/repository
export CCNR_DIRECTORY=$HOME/.ccnx/repository
export CCNR_GLOBAL_PREFIX
ccnr 2>>$CCNR_DIRECTORY/log &
fi

